File: mobile/packages/native_video_editor/android/src/main/kotlin/io/ente/native_video_editor/core/VideoModels.kt
package io.ente.native_video_editor.core

import kotlin.math.max
import kotlin.math.roundToInt

data class VideoMetadata(
    val durationMs: Long,
    val width: Int,
    val height: Int,
    val rotationDegrees: Int,
    val bitrate: Long?,
    val frameRate: Float?
) {
    val displayWidth: Int
        get() = if (rotationDegrees == 90 || rotationDegrees == 270) height else width

    val displayHeight: Int
        get() = if (rotationDegrees == 90 || rotationDegrees == 270) width else height
}

enum class VideoFramePolicy {
    PRECISE,
    NEAREST_SYNC;
}

data class VideoFrameRequest(
    val inputPath: String,
    val outputPaths: List<String>,
    val positionsMs: List<Long>,
    val maxWidth: Int,
    val maxHeight: Int,
    val quality: Int,
    val policy: VideoFramePolicy
) {
    init {
        require(inputPath.isNotBlank()) { "Missing input path" }
        require(outputPaths.isNotEmpty() && outputPaths.size == positionsMs.size) {
            "Every timestamp requires an output path"
        }
        require(outputPaths.none(String::isBlank) && outputPaths.distinct().size == outputPaths.size) {
            "Output paths must be non-empty and unique"
        }
        require(positionsMs.all { it >= 0 }) { "Frame timestamps must be non-negative" }
        require(maxWidth > 0 && maxHeight > 0) { "Frame bounds must be positive" }
        require(quality in 1..100) { "JPEG quality must be between 1 and 100" }
    }
}

data class VideoFrameResult(
    val outputPath: String,
    val width: Int,
    val height: Int
)

data class VideoFrameExtractionResult(
    val videoInfo: VideoMetadata,
    val frames: List<VideoFrameResult>
)

data class VideoCrop(
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int
) {
    init {
        require(x >= 0 && y >= 0 && width > 0 && height > 0) {
            "Invalid crop rectangle"
        }
    }
}

data class VideoEditRequest(
    val inputPath: String,
    val outputPath: String,
    val trimStartMs: Long?,
    val trimEndMs: Long?,
    val rotateDegrees: Int?,
    val crop: VideoCrop?
) {
    init {
        require(inputPath.isNotBlank() && outputPath.isNotBlank()) { "Missing video path" }
        require((trimStartMs == null) == (trimEndMs == null)) {
            "Trim bounds must be provided together"
        }
        require(trimStartMs == null || (trimStartMs >= 0 && trimStartMs < trimEndMs!!)) {
            "Invalid trim bounds"
        }
        require(rotateDegrees == null || rotateDegrees in setOf(0, 90, 180, 270)) {
            "Invalid rotation"
        }
    }
}

data class VideoEditResult(val outputPath: String, val isReEncoded: Boolean)

internal data class PixelSize(val width: Int, val height: Int)

internal fun boundedVideoSize(
    width: Int,
    height: Int,
    maxWidth: Int,
    maxHeight: Int
): PixelSize {
    require(width > 0 && height > 0) { "Video dimensions are unavailable" }
    val scale = minOf(
        1.0,
        maxWidth.toDouble() / width,
        maxHeight.toDouble() / height
    )
    return PixelSize(
        max(1, (width * scale).roundToInt()),
        max(1, (height * scale).roundToInt())
    )
}

internal fun clampFramePositionMs(positionMs: Long, durationMs: Long): Long {
    if (durationMs <= 0) return max(0L, positionMs)
    return positionMs.coerceIn(0L, max(0L, durationMs - 1L))
}

File: mobile/packages/native_video_editor/ios/Classes/VideoModels.swift
import AVFoundation

public struct VideoMetadata: Equatable, Sendable {
    public let durationMs: Int64
    public let width: Int
    public let height: Int
    public let displayWidth: Int
    public let displayHeight: Int
    public let rotationDegrees: Int
    public let bitrate: Int64?
    public let frameRate: Double?

    public init(
        durationMs: Int64,
        width: Int,
        height: Int,
        displayWidth: Int,
        displayHeight: Int,
        rotationDegrees: Int,
        bitrate: Int64?,
        frameRate: Double?
    ) {
        self.durationMs = durationMs
        self.width = width
        self.height = height
        self.displayWidth = displayWidth
        self.displayHeight = displayHeight
        self.rotationDegrees = rotationDegrees
        self.bitrate = bitrate
        self.frameRate = frameRate
    }
}

public enum VideoFramePolicy: Sendable {
    case precise
    case nearestSync
}

public struct VideoFrameRequest: Sendable {
    public let inputURL: URL
    public let outputURLs: [URL]
    public let positionsMs: [Int64]
    public let maxWidth: Int
    public let maxHeight: Int
    public let quality: Int
    public let policy: VideoFramePolicy

    public init(
        inputURL: URL,
        outputURLs: [URL],
        positionsMs: [Int64],
        maxWidth: Int,
        maxHeight: Int,
        quality: Int,
        policy: VideoFramePolicy
    ) throws {
        let inputURL = inputURL.standardizedFileURL.resolvingSymlinksInPath()
        let outputURLs = outputURLs.map {
            $0.standardizedFileURL.resolvingSymlinksInPath()
        }
        guard FileManager.default.fileExists(atPath: inputURL.path),
            !positionsMs.isEmpty,
            positionsMs.count == outputURLs.count,
            positionsMs.allSatisfy({ $0 >= 0 }),
            !outputURLs.contains(inputURL),
            Set(outputURLs.map(\.path)).count == outputURLs.count,
            maxWidth > 0,
            maxHeight > 0,
            (1...100).contains(quality)
        else {
            throw VideoEditorError.invalidRequest("Invalid frame extraction request")
        }
        self.inputURL = inputURL
        self.outputURLs = outputURLs
        self.positionsMs = positionsMs
        self.maxWidth = maxWidth
        self.maxHeight = maxHeight
        self.quality = quality
        self.policy = policy
    }
}

public struct VideoFrameResult: Equatable, Sendable {
    public let outputURL: URL
    public let width: Int
    public let height: Int
}

public struct VideoFrameExtractionResult: Equatable, Sendable {
    public let videoInfo: VideoMetadata
    public let frames: [VideoFrameResult]
}

public struct VideoCrop: Equatable, Sendable {
    public let x: Int
    public let y: Int
    public let width: Int
    public let height: Int

    public init(x: Int, y: Int, width: Int, height: Int) throws {
        guard x >= 0, y >= 0, width > 0, height > 0 else {
            throw VideoEditorError.invalidRequest("Invalid crop rectangle")
        }
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

public struct VideoEditRequest: Sendable {
    public let inputURL: URL
    public let outputURL: URL
    public let trimStartMs: Int64?
    public let trimEndMs: Int64?
    public let rotateDegrees: Int?
    public let crop: VideoCrop?

    public init(
        inputURL: URL,
        outputURL: URL,
        trimStartMs: Int64?,
        trimEndMs: Int64?,
        rotateDegrees: Int?,
        crop: VideoCrop?
    ) throws {
        let inputURL = inputURL.standardizedFileURL.resolvingSymlinksInPath()
        let outputURL = outputURL.standardizedFileURL.resolvingSymlinksInPath()
        guard inputURL != outputURL,
            FileManager.default.fileExists(atPath: inputURL.path),
            (trimStartMs == nil) == (trimEndMs == nil),
            trimStartMs == nil || (trimStartMs! >= 0 && trimStartMs! < trimEndMs!),
            rotateDegrees == nil || [0, 90, 180, 270].contains(rotateDegrees!)
        else {
            throw VideoEditorError.invalidRequest("Invalid video processing request")
        }
        self.inputURL = inputURL
        self.outputURL = outputURL
        self.trimStartMs = trimStartMs
        self.trimEndMs = trimEndMs
        self.rotateDegrees = rotateDegrees
        self.crop = crop
    }
}

public struct VideoEditResult: Equatable, Sendable {
    public let outputURL: URL
    public let isReEncoded: Bool
}

public enum VideoEditorError: LocalizedError {
    case invalidRequest(String)
    case metadataUnavailable(String, Error? = nil)
    case frameExtraction(String, Error? = nil)
    case frameWrite(String, Error? = nil)
    case invalidTimeRange(String)
    case composition(String, Error? = nil)
    case export(String, Error? = nil)

    public var message: String {
        switch self {
        case .invalidRequest(let message),
            .metadataUnavailable(let message, _),
            .frameExtraction(let message, _),
            .frameWrite(let message, _),
            .invalidTimeRange(let message),
            .composition(let message, _),
            .export(let message, _):
            return message
        }
    }

    public var errorDescription: String? { message }

    public var underlyingError: Error? {
        switch self {
        case .metadataUnavailable(_, let error),
            .frameExtraction(_, let error),
            .frameWrite(_, let error),
            .composition(_, let error),
            .export(_, let error):
            return error
        case .invalidRequest, .invalidTimeRange:
            return nil
        }
    }
}
